sklearn/tests/test_base.py

```diff
@@ -359,3 +357,81 @@ def test_pickle_version_warning():
     # check that no warning is raised for external estimators
     TreeNoVersion.__module__ = "notsklearn"
     assert_no_warnings(pickle.loads, tree_pickle_noversion)
+
+
+class DontPickleCacheMixin(object):
+    def __getstate__(self):
+        data = self.__dict__.copy()
+        data["_cache"] = None
+        return data
+
+
+class MultiInheritanceEstimator(BaseEstimator, DontPickleCacheMixin):
+    def __init__(self, b=5):
+        self.b = b
+        self._cache = None
+
+    @property
+    def cache(self):
+        if not self._cache:
+            self._cache = "some value"
+        return self._cache
+
+
+class SingleInheritanceEstimator(BaseEstimator):
+    def __init__(self, b=5):
+        self.b = b
+        self._cache = None
+
+    def __getstate__(self):
+        data = self.__dict__.copy()
+        data["_cache"] = None
+        return data
+
+    @property
+    def cache(self):
+        if not self._cache:
+            self._cache = "some value"
+        return self._cache
+
+
+class TestPicklingConstraints(object):
```

Comment:
1. We usually don't use test classes, but I'm personally okay with this as a way of grouping the code.
2. It might be good to mention that this test is about `BaseEstimator` somewhere (though I see there's a lack of that in this file).